proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,222 @@
+
+We propose to also store the new WALs in separate directories e.g. `wal.v2`. Thanks to that the rewrite from one version to another is eventual and can be done segment by segment.
+The additional advantage of this way of versioning is that it's clear when your WAL fully migrated to a certain version.
+
+Finally, we propose that all Prometheus releases will contain the following table:
+
+| Data        | Supported | Writes |
+|-------------|-----------|--------|
+| WAL         | 1, 2      | 1      |
+| Block index | 2         | 2      |
+
+```
+The last revertable Prometheus version: v3.0 
+```
+
+See [alternatives](#alternatives) for other ideas.
+
+### Two-Fold Migration Strategy
+
+Given the following example:
+
+![twofold.png](../assets/2024-11-25_changing_wal_format/twofold.png)
+
+1. We release Prometheus X+1 version that supports both Y and Y+1 data but still writes Y.
+2. We release Prometheus X+2 version that supports both Y and Y+1 data, but now it writes new data as Y.
+
+Pros:
+* Users have a durable rollout path back and forth.
+* Dev has clarity on how to develop "breaking revert" changes.
+* In theory, it allows revert to X from X+2, by going through X+1 and ensuring all data was migrated (eventually) to Y version. In practice however that eventuality is long, or hard to discover.
+
+Cons:
+* It can catch users by surprise
+  * **Mitigation**: We also plan guide and documentation.
+* It takes time to rollout new changes if you want them fast.
+  * **Mitigation**: We plan to add a flag to opt-in sooner
+* It is a breaking change which shouldn't be made in a major release.
+* **Mitigation**: We accept that fact, given precedence and no other way to support both formats in the same time without major performance penalties.
+
+## Alternatives
+
+### Require LTS support
+
+We could add a variation to the [Two-Fold Migration Strategy](#two-fold-migration-strategy) where both X+1 and the last LTS (long time support version) before X+1, should understand Y+1 version -- before X+2 can write it by default.
+
+Pros:
+* This gives a bit more stability to users and less surprise.
+
+Cons:
+* We don't do LTS-es regularly
+* Is it safe or feasible to add Y+1 support to the old LTS code?
+
+### Add a CLI tool that rewrites WAL to a specific version
+
+We could add a `promtool` command that rewrites WAL segments to a given version.
+
+Users can then migrate their WAL with a single command either as an init process before reverting or manually on a disk.
+
+This should take ~minutes even for larger WAL files.
+
+Pros:
+* Less need for a two-fold strategy?
+
+Cons:
+* We need to write migration code and ensure it's efficient enough for bigger data
+* A bit painful to use on scale and remotely (e.g. on Kubernetes)
+
+### Rewrite WAL before/during replay
+
+Instead of supporting multiple directories for WAL for various versions, we could rewrite WAL on the start.
```

Comment:
Cons+ I think the risk is that the rewrite fails and we cannot go back , resulting in data loss. Read only WAL on startup seems safer.